pallets/proofs-dealer/src/utils.rs

```diff
@@ -1,25 +1,51 @@
+// TODO: Remove this attribute once the file is implemented.
 #![allow(dead_code)]
 #![allow(unused_variables)]
-use frame_support::{ensure, pallet_prelude::DispatchResult};
+use frame_support::{
+    ensure,
+    pallet_prelude::DispatchResult,
+    traits::{fungible::Mutate, tokens::Preservation, Get},
+};
 use frame_system::pallet_prelude::BlockNumberFor;
 use sp_trie::CompactProof;
 use storage_hub_traits::ProvidersInterface;
 
 use crate::{
     pallet,
-    types::{AccountIdFor, BalanceFor, FileKeyFor, ProviderFor, ProvidersPalletFor},
+    types::{
+        AccountIdFor, BalanceFor, BalancePalletFor, ChallengesFeeFor, FileKeyFor, ProviderFor,
+        ProvidersPalletFor, TreasuryAccountFor,
+    },
     ChallengesQueue, Error, Pallet,
 };
 
 impl<T> Pallet<T>
 where
     T: pallet::Config,
 {
-    // TODO: Document.
+    /// Add custom challenge to ChallengesQueue.
+    ///
+    /// Check if sender is a registered Provider. If it is not, charge a fee for the challenge.
+    /// This is to prevent spamming the network with challenges. If the challenge is already queued,
+    /// just return. Otherwise, add the challenge to the queue.
```

Comment:
I am not sure I understand the reason behind this. 
This opens up the possibility for a storage provider to stall the network (intentionally or unintentionally).

Reply:
It doesn't really stall it, it just fills it up with challenges that they themselves are gonna have to answer to. And in any case, it goes to the regular challenges queue, not the priority, which is the one used for deletion requests and so on. I have a todo somewhere in the extrinsic that uses this about considering only allowing free challenges when there has been a change of an MSP, but that would introduce a dependency of this pallet with the File System pallet, that I'd like to avoid now.

Reply:
I like the last comment about specific scenarios when it is free. I don’t think it should be free all the time. But we can discuss about this another time.